packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4585,652 +4352,1041 @@ mixin RelayoutWhenSystemFontsChangeMixin on RenderObject {
-          // The fragment is not explicit, its elevation has been absorbed by
-          // the parent config (as thickness). We still need to make sure that
-          // its children are placed at the elevation dictated by this config.
-          elevationAdjustment: elevationAdjustment + _config.elevation,
-          result: result,
-          siblingNodes: siblingNodes,
-        );
       }
+    }
+  }
+
+  /// Ensures the semantics nodes from this render object semantics subtree are
+  /// generated and up to date.
+  ///
+  /// This can only be called if the owning rendering object is a semantics
+  /// boundary. For non boundary rendering objects, they require semantics
+  /// information from both their parent and child rendering objects to update
+  /// its cache, so it can't update by themselves.
+  void ensureSemantics() {
+    assert(configProvider.effective.isSemanticBoundary || isRoot);
+    if (isRoot) {
+      buildSemantics(
+        usedSemanticsIds: <int>{},
+        geometry: _SemanticsGeometry.root,
+      );
+    } else {
+      assert(built);
+      // parent data and parent geometry don't change, there isn't anything to
+      // update for semantics nodes generated in this render object semantics.
+      buildSemanticsSubtree(
+        geometry: _SemanticsGeometry(
+          transform: Matrix4.identity(),
+          paintClipRect: cachedSemanticsNode!.parentPaintClipRect,
+          semanticsClipRect: cachedSemanticsNode!.parentSemanticsClipRect,
+        ),
+        usedSemanticsIds: <int>{},
+        elevationAdjustment: 0.0,
+      );
+    }
+
+    // assert(() {
+    //   // Must be formed a semantics node.
+    //   if (!_formedSemanticsNode!) {
+    //     return false;
+    //   }
+    //   final SemanticsNode producedSemanticsNode = cachedSemanticsNode!;
+    //   return semanticsNodes.any((SemanticsNode node) => node == producedSemanticsNode);
+    // }());
+  }
+
+  void didUpdateParentData(_SemanticsParentData newParentData) {
+    if (parentData == newParentData) {
       return;
     }
+    // Parent data changes may result in node formation changes.
+    markNeedsBuild();
+    // print('renderobject sets new parent data ${describeIdentity(renderObject)}');
+    parentData = newParentData;
+    // var children = <String>[];
+    // renderObject.visitChildrenForSemantics((RenderObject renderChild) {
+    //   children.add(describeIdentity(renderChild));
+    // });
+    // print('didUpdateParentData for ${describeIdentity(renderObject)}, children $children');
+    updateChildParentData();
+
+    final Set<SemanticsTag>? tags = newParentData.tagsForChildren;
+    if (tags != null) {
+      assert(tags.isNotEmpty);
+      configProvider.setProperty((SemanticsConfiguration config) {
+        tags.forEach(config.addTagForChildren);
+      });
+    }
+
+    final bool effectiveBlocksUserAction = newParentData.blocksUserActions || configProvider.original.isBlockingUserActions;
```

Comment:
Related to the `original` vs `effective` comment:

Could you clarify why we are getting the effective "block user action" from `original` but then comparing it with the one in `effective`?

My understanding so far is that `effective` is either identical to `original` OR a writable copy of `original`. And by doing `setProperty` below, you are basically making sure that `effective` is a writable copy of `original` and then you update it.

If my understanding is correct, then I think it makes more sense to do:
```suggestion
    final bool effectiveBlocksUserAction = newParentData.blocksUserActions || configProvider.effective.isBlockingUserActions;
```

Reply:
The entire class is private so I dont think it really matter, but I try to maintain certain code style that property used outside of this class is public while internal property stay private

Reply:
also I think `mutable` doesn't convey the meaning well, it state the what happen to it but is not helpful that when it should be used.